frontend/src/components/ui/data-grid/data-grid.ts

```diff
@@ -0,0 +1,320 @@
+import { localized, msg } from "@lit/localize";
+import type { SlChangeEvent, SlInput } from "@shoelace-style/shoelace";
+import clsx from "clsx";
+import { css, html, nothing } from "lit";
+import { customElement, property } from "lit/decorators.js";
+import { ifDefined } from "lit/directives/if-defined.js";
+import { repeat } from "lit/directives/repeat.js";
+import { nanoid } from "nanoid";
+import type { EmptyObject } from "type-fest";
+
+import { DataGridRowsController } from "./controllers/rows";
+import type { DataGridRow, RowRemoveEventDetail } from "./data-grid-row";
+import type { GridColumn, GridItem } from "./types";
+
+import { TailwindElement } from "@/classes/TailwindElement";
+import { pluralOf } from "@/utils/pluralize";
+import { tw } from "@/utils/tailwind";
+
+/**
+ * Data grids structure data into rows and and columns.
+ *
+ * @slot label
+ * @slot rows
+ * @fires btrix-change
+ * @fires btrix-remove
+ */
+@customElement("btrix-data-grid")
+@localized()
+export class DataGrid extends TailwindElement {
+  static styles = css`
+    :host {
+      --border: 1px solid var(--sl-panel-border-color);
+    }
+
+    btrix-data-grid-row:not(:first-of-type),
+    btrix-table-body ::slotted(*:nth-of-type(n + 2)) {
+      border-top: var(--border) !important;
+    }
+
+    btrix-data-grid-row,
+    btrix-table-body ::slotted(btrix-data-grid-row) {
+      /* TODO Support different input sizes */
+      min-height: calc(var(--sl-input-height-medium) + 1px);
+    }
+  `;
+
+  /**
+   * Set of columns.
+   */
+  @property({ type: Array })
+  columns?: GridColumn[];
+
+  /**
+   * Set of data to be presented as rows. Omit if using the `rows` slot.
+   */
+  @property({ type: Array })
+  items?: GridItem[];
+
+  /**
+   * Stick header row to the top of the viewport.
+   */
+  @property({ type: Boolean })
+  stickyHeader = false;
+
+  /**
+   * Item key to use as the row key, like an ID.
+   * Defaults to one generated by nanoid.
+   */
+  @property({ type: String })
+  rowKey?: string;
+
+  /**
+   * Whether rows can be removed.
+   */
+  @property({ type: Boolean })
+  removeRows = false;
+
+  /**
+   * Whether rows can be added.
+   */
+  @property({ type: Boolean })
+  addRows = false;
+
+  /**
+   * Make the number of rows being added configurable,
+   * with a default starting value.
+   */
+  @property({ type: Number })
+  addRowsInputValue?: number;
+
+  /**
+   * Whether cells can be edited.
+   */
+  @property({ type: Boolean })
+  editCells = false;
+
+  /**
+   * Disable an editable grid.
+   */
+  @property({ type: Boolean })
+  disabled?: boolean;
+
+  /**
+   * Default item for new rows.
+   */
+  @property({ type: Object })
+  defaultItem?: EmptyObject | GridItem = {};
+
+  /**
+   * Text for form control label. Use slot to include markup.
+   */
+  @property({ type: String })
+  formControlLabel?: string;
+
+  /**
+   * ID for form control label.
+   */
+  @property({ type: String })
+  formControlLabelId = nanoid();
```

Comment:
While not _technically_ disallowed, ids generally shouldn't start with numbers
```suggestion
  formControlLabelId = `label-${nanoid()}`;
```